### DISCUSSION TITLE
error uniform.type == GL_SAMPLER_2D or GL_SAMPLER_CUBE after shader compile

### DISCUSSION THREAD

**Discussion by 10100010010001001010** [axmolengine/axmol#2250] [Q&A] [answered]
my fragment shader:
```
#version 300 es
precision mediump float;

layout(std140) uniform fs_ub {
    highp vec2 u_lightPos;
    highp float u_lightRad; 
    highp float u_darkness;     
} _15;

in vec4 v_position;
out vec4 FragColor;

void main()
{
    vec2 screenPos = v_position.xy * 0.5 + 0.5;

    float dist = distance(screenPos, _15.u_lightPos);

    float darknessFactor = smoothstep(_15.u_lightRad, _15.u_lightRad * 1.7, dist);

    float alpha = darknessFactor * _15.u_darkness;

    FragColor = vec4(0.0, 0.0, 0.0, alpha);
}
```
i want realise a simple darkness effect, but I get an error from title of topic. This error from string 392 of "ProgramGL.cpp" file, and i see a hint "must be sampler: sampler2D or samplerCube". But i dont use a texture, just vertices, like the shader from the tests called "quadColor_fs", the structure of which i was guided by.
Please help

**Comment by rh101** [comment]
You seem to have deleted lines from `{axmol}/core/renderer/shaders/quadColor.frag` that are actually required.  Also, the `quadColor_fs` is the output shader, which you shouldn't be copying.  Look at the original source at `{axmol}/core/renderer/shaders/quadColor.frag` etc..

Try something like this:
```
#version 310 es
precision mediump float;

layout(location = COLOR0) in vec4 v_color;
layout(location = TEXCOORD0) in vec2 v_texCoord;
layout(location = SV_Target0) out vec4 FragColor;

layout(binding = 0) uniform sampler2D u_tex0;

layout(std140) uniform fs_ub {
    highp vec2 u_lightPos;
    highp float u_lightRad; 
    highp float u_darkness;  
};

void main()
{
    vec2 screenPos = v_texCoord.xy * 0.5 + 0.5;

    float dist = distance(screenPos, u_lightPos);

    float darknessFactor = smoothstep(u_lightRad, u_lightRad * 1.7, dist);

    float alpha = darknessFactor * u_darkness;

    FragColor = vec4(0.0, 0.0, 0.0, alpha);
}
```

**Reply by 10100010010001001010** [reply]
Forgot to mention that I tried specifying sampler2d, now I just have a fragment shader compilation error

**Reply by rh101** [reply]
Change the version line to `#version 310 es`.  The shader code from my previous post successfully compiled when the version is changed to 310.

**Reply by 10100010010001001010** [reply]
Same result - compiling error

**Reply by 10100010010001001010** [reply]
Any other ideas what the problem could be?

**Reply by rh101** [reply]
Which platform are you building for?  The test I did was for a Windows build.

Have you tried this with a new project? If not, then generate a new axmol project, add the shader to it, then attempt to build it. If it doesn't work, then please share the project on here.

**Reply by 10100010010001001010** [reply]
I'm building for Win10, and I created a new project on the updated version of axmol, I still get the same error. Here my project below, I think there are bugs in my class, but compilation stops on load program at line 14 from ShaderNode.cpp.
[test.zip](https://github.com/user-attachments/files/17932456/test.zip)

**Reply by halx99** [reply]
your vertex shader is incorrect, all uniforms except sampler should in block, please modify

```#version 310 es
precision highp float;

uniform mat4 u_MVPMatrix;

layout(location = 0) in vec4 a_position;
out vec4 v_position;

void main()
{
    gl_Position = u_MVPMatrix * a_position;
    v_position = a_position;
}
``` 
to 
```
#version 310 es
precision highp float;

layout(std140) uniform vs_ub {
    uniform mat4 u_MVPMatrix;
};

layout(location = 0) in vec4 a_position;
out vec4 v_position;

void main()
{
    gl_Position = u_MVPMatrix * a_position;
    v_position = a_position;
}
```

**Reply by 10100010010001001010** [reply]
I tried this option, I still get a compilation error, I thought to look in the console, it says this:
![image](https://github.com/user-attachments/assets/ae97c468-0beb-401b-b490-9e6bda568150)

**Reply by halx99** [reply]
as @rh101 said, don't load it directly, instead, read the shader workflow of axmol2:  https://github.com/axmolengine/axmol/issues/1279

**Reply by rh101** [reply]
@10100010010001001010 The shader files must be located in `{yourproject}/Source/shaders/`, and not in `Content/shaders/`.  When you build the application, the shaders files will be processed, and the correct files for the target will be generated.  

In your case, with files named "vertex_shader.vsh" and "fragment_shader.fsh", two new files will be generated, named "vertex_shader_vs" and "fragment_shader_fs".  In your code, you need to load these generated files, and not your original shaders.  To reference them, you need to use the following:

`custom/vertex_shader_vs`
and
`custom/fragment_shader_fs`

Custom shaders included in projects are always prefixed with `custom/`.

**Reply by rh101** [reply, marked as answer]
Here are the fixed shaders based on your project and the original fragment shader you posted, where I accidentally removed `v_position`:

`vertext_shader.vsh`:

```
#version 310 es

layout(std140) uniform vs_ub {
    mat4 u_MVPMatrix;
};

layout(location = POSITION) in vec4 a_position;
layout(location = 0) out vec4 v_position;

void main()
{
    gl_Position = u_MVPMatrix * a_position;
    v_position = a_position;
}
```

and `fragment_shader.fsh` (with v_position re-added):

```
#version 310 es
precision mediump float;

layout(location = COLOR0) in vec4 v_color;
layout(location = POSITION) in vec4 v_position;
layout(location = SV_Target0) out vec4 FragColor;

layout(binding = 0) uniform sampler2D u_tex0;

layout(std140) uniform fs_ub {
    highp vec2 u_lightPos;
    highp float u_lightRad; 
    highp float u_darkness;  
};

void main()
{
    vec2 screenPos = v_position.xy * 0.5 + 0.5;

    float dist = distance(screenPos, u_lightPos);

    float darknessFactor = smoothstep(u_lightRad, u_lightRad * 1.7, dist);

    float alpha = darknessFactor * u_darkness;

    FragColor = vec4(0.0, 0.0, 0.0, alpha);
}

```

**Reply by 10100010010001001010** [reply]
Thank you very much for the detailed explanation

**Comment by rh101** [comment]
I grabbed the info from #1279 and added it to the wiki [here](https://github.com/axmolengine/axmol/wiki/Shaders-in-Axmol), since these shader questions seem to be coming up more often.
